### PR TITLE
[RFC] denc: add rvalue overloads for bufferlist/ptr encode

### PR DESCRIPTION
**Pull request by cbodley**
adds denc() and ceph::encode() overloads taking `bufferlist&&` and `bufferptr&&` so that buffer segments can be moved directly into the destination bufferlist without the atomic reference counting associated with sharing the buffers between the source and destination

this is mainly only useful for manual encoding, though composite types (like map<string, bufferlist> or structs that contain buffers) could also add rvalue overloads to take advantage of this

### CODE REVIEW

File: src/include/encoding.h [hunk 1]
```diff
@@ -222,13 +222,20 @@ inline void encode(const char *s, bufferlist& bl)
 // buffers
 
 // bufferptr (encapsulated)
-inline void encode(const buffer::ptr& bp, bufferlist& bl) 
+inline void encode(const buffer::ptr& bp, bufferlist& bl)
```
Comment (rzarzynski) [inline src/include/encoding.h]: Thanks for cleaning up the whitespaces.

File: src/include/buffer.h [hunk 2]
```diff
@@ -585,6 +585,20 @@ namespace buffer CEPH_BUFFER_API {
...
+	if (!p.length()) {
+	  return;
+	}
+	if (deep) {
+	  append(p.c_str(), p.length());
+          p = bufferptr{}; // drop ref
+	} else {
+	  flush_and_continue();
+	  out_of_band_offset += p.length();
+	  pbl->append(std::move(p));
```
Comment (rzarzynski) [inline src/include/buffer.h]: What about using `buffer::list::push_back(ptr&&)` instead?  It's defined in `buffer.h` so compilers would have a chance to inline the code.

File: src/include/buffer.h [hunk 3]
```diff
@@ -599,6 +613,21 @@ namespace buffer CEPH_BUFFER_API {
...
+	}
+	if (deep) {
+	  for (const auto &p : l._buffers) {
+	    append(p.c_str(), p.length());
+	  }
+	  l.clear();
+	} else {
+	  flush_and_continue();
+	  out_of_band_offset += l.length();
+	  pbl->claim_append(l);
```
Comment (rzarzynski) [inline src/include/buffer.h]: This is cool.

File: src/include/denc.h [hunk 4]
```diff
@@ -763,6 +763,12 @@ struct denc_traits<bufferptr> {
     p.append(v);
   }
   template <class It>
+  static std::enable_if_t<!is_const_iterator_v<It>>
+  encode(bufferptr&& v, It& p, uint64_t f=0) {
+    denc((uint32_t)v.length(), p);
+    p.append(std::move(v));
+  }
+  template <class It>
```
Comment (tchaikov) [inline src/include/denc.h]: probably we could unify this variant with the `bufferptr&` one by templating them and perfect forwarding.  maybe this approach could also pave the road to the dencoder for `seastar::temporary_buffer`.

**Comment by cbodley** [comment]
thanks for review!

> this is mainly only useful for manual encoding, though composite types (like map<string, bufferlist> or structs that contain buffers) could also add rvalue overloads to take advantage of this

can you guys think of other places that can take advantage of this?

**Comment by cbodley** [comment]
closing until we find other uses for this

**Pull request closed without merging**